Autopilot/Inc/SensorFusion.hpp

```diff
@@ -0,0 +1,75 @@
+/**
+ * Sensor Fusion - Converts raw sensor data into human readable formatted structs
+ * Author: Lucy Gong, Dhruv Rawat, Anthony Bertnyk
+ */
+
+#include "AttitudeDatatypes.hpp"
+
+#ifndef SENSORFUSION_HPP
+#define SENSORFUSION_HPP
+
+// -1 = FAILED
+// 0 = SUCCESS
+// 1 = Old Data
+struct SFError_t{
+    int errorCode;
+};
+
+struct SFOutput_t {
+    float roll, pitch, yaw; //rad
+    float rollRate, pitchRate, yawRate; //rad/s
+    float airspeed; //m/s
+    float altitude; //m
+    float altitudeSpeed; //m/s
+    long double latitude; //Decimal degrees
+    float latitudeSpeed; //m/s
+    long double longitude; //Decimal degrees
+    float longitudeSpeed; //m/s
+};
+
+/**
+ * Initiate sensor fusion.
+ */ 
+void SF_Init(void);
+
+/**
+ * Get fused sensor data.
+ * @param[out]    output  Reference to an output struct for fused data.
+ */ 
+SFError_t SF_GetResult(SFOutput_t *output);
+
+//RTOS stuff that the cpp files need
+#ifdef TARGET_BUILD
+
+extern "C"
+{
+#include "cmsis_os.h"
+}
+
+const char SF_MAIL_Q_SIZE = 1;
+
+//Set up a mail queue for getting data from sensor fusion
+extern osMailQDef(SFMailQ, SF_MAIL_Q_SIZE, SFOutput_t);
```

Comment:
You don't actually need a queue here. Your module will be called by multiple tasks, yes, but in your case, there is a far easier way to share data. Each task in FreeRtos gets its own stack, but the heap is shared by all tasks, and since FreeRtos doesen't deal with virtual memory (it's all physical), then static data is also shared by all tasks.

What this means to you is that if you declare a  static or global variable in your .cpp file, it will be shared by all tasks (because statics/globals live in static memory). A variable declared on the heap would also be able to be shared (but that's not as clean of a solution cuz you gotta pass pointers around).

Reply:
Would that prevent multiple tasks from accessing the output struct at once? Maybe a queue is overkill, but I am concerned that sensor fusion would start writing to the static/global variable, but partway through another task tries to access it and receives incomplete/broken data.

Reply:
The way to prevent that is with a Mutex. Mutexes ensure that only one thread can access the ressource at once. 

https://www.freertos.org/Real-time-embedded-RTOS-mutexes.html#:~:text=accompanying%20example%20projects.-,FreeRTOS%20Mutexes,include%20a%20priority%20inheritance%20mechanism.&text=When%20used%20for%20mutual%20exclusion,('take')%20the%20token.

You might like to look deeper into this as well, cuz one optimization is that a thread that's reading doesen't need to wait for another thread that's reading, only for a thread that's writing, so there may be a way to have all the reading threads to all share the token somehow.

This should all live in the cpp file though.